### ISSUE TITLE
whatwg-fetch and isomorphic-fetch should include Error type declarations.

### ISSUE THREAD

**Issue by ty10r** [DefinitelyTyped/DefinitelyTyped#12460]
- [x] I tried using the latest `isomorphic-fetch/index.d.ts` file in this repo and had problems.
- [x] I tried using the latest `whatwg-fetch/index.d.ts` file in this repo and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] I want to talk about `whatwg-fetch/index.d.ts`.
  - The authors of that type definition are cc/ @ryan-codingintrigue
- [x] I want to talk about `isomorphic-fetch/index.d.ts`.
  - The authors of that type definition are cc/ @toddlucas

I understand that these are technically separate type declarations, but I imagine that their error types should be aligned with the whatwg fetch standard.

I believe that it would be worth including error types in the type declarations of both of these modules. Otherwise, it forces the consumer of these modules to all create a copy of the error types or treat them as a weak `any` type and translate to another error type domain.

**Comment by ty10r** [comment]
I do not have permission, but I believe this issue should be labeled as `@types`

**Comment by ryan-codingintrigue** [comment]
Which error types are you referring to? The rejected promise type?

**Comment by ty10r** [comment]
@ryan-codingintrigue Yes, while using `isomorphic-fetch` I was forcing a call to fetch to reject with a network error.
In one case, I was receiving an error:
```js
{
  name: 'FetchError',
  message: 'request to https:***** failed, reason: getaddrinfo ENOTFOUND ******',
  type: 'system',
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND' 
}
```

I expected to find some interface for `FetchError` in the type declarations, but did not find any Error interfaces. So I checked whatwg-fetch and found the same issue there.

**Comment by ty10r** [comment]
@ryan-codingintrigue I just read the `whatwg-fetch` code and believe that I may not have been correct by including it in this issue. All of the rejection errors in that code are `TypeError`, so I'm not entirely right to expect an interface or error type in that declaration file. 

Am I wrong to think that it could be worth having a return value of something like `Promise<Response | TypeError>` for fetch's signature?

**Comment by ryan-codingintrigue** [comment]
@ty10r The definition for the Promise interface actually only defines a *resolution* type:
https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts

And the `onRejected` callback is unfortunately hard-coded to by `any` in the definition:
https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L62

tldr; It's a limitation of the TS Promise definition. More discussion here: https://github.com/Microsoft/TypeScript/issues/7588

**Comment by ty10r** [comment]
@ryan-codingintrigue Thanks for the info. I'll close this for now.
